### DISCUSSION TITLE
Can I compare two float vectors with an expected accuracy?

### DISCUSSION THREAD

**Discussion by jarenduan** [mathnet/mathnet-numerics#814] [Q&A] [answered]
The Vector.Equals methods implementations are doing precisely comparison for float numbers. However, in many real-world calculations, we only care about the equality under an expected accuracy. For example, if I only care about an accuracy at scale of 0.001, so I wish the two following vectors are equal to each other:
    v1: <1.00001, 2.000001>
    v2: <1.0 , 2.0>
.
In this case, in what way I can directly compare v1 and v2 to get the v1==v2?  Should we add a overload to Equals method?

**Comment by cdrnet** [comment, marked as answer]
You can use `AlmostEqual` for that effect:

```
var v1 = DenseVector.OfArray(new[] {100000.00001, 200000.000001 });
var v2 = DenseVector.OfArray(new[] {100000.0, 200000.0 });

v1.Equals(v2); // false
v1.AlmostEqual(v2, 0.001d); // true (delta)
v1.AlmostEqual(v2, 0.000001d); // false (delta)
v1.AlmostEqual(v2, 3); // true (decimal places)
v1.AlmostEqual(v2, 7); // false (decimal places)
v1.AlmostEqualRelative(v2, 0.001d); // true (delta)
v1.AlmostEqualRelative(v2, 0.000001d); // true (delta)
v1.AlmostEqualRelative(v2, 0.00000000001d); // false (delta)
v1.AlmostEqualRelative(v2, 3); // true (decimal places)
v1.AlmostEqualRelative(v2, 7); // true (decimal places)
v1.AlmostEqualRelative(v2, 10); // false (decimal places)
```

Is this what you're looking for?

**Comment by jarenduan** [comment]
@cdrnet Yes, thanks, and that's exactly what I want. 

Maybe it would be better to make them a overload of Equals() methods, but that's another issue.

**Comment by cdrnet** [comment]
With an enum as optional 3rd parameter to choose between absolute (default) and relative?

**Comment by 23W** [comment]
@cdrnet 
I don't understand difference between `AlmostEqualRelative`  and  `AlmostEqual`. Can you please clarify. 
`AlmostEqualRelative`  is relative variant but it's relative to what?
